core/unit_test/TestRangePolicyConstructors.hpp

```diff
@@ -0,0 +1,391 @@
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#include <cstdio>
+#include <sstream>
+#include <iostream>
+
+#include <gtest/gtest.h>
+
+#include <Kokkos_Core.hpp>
+
+namespace Test {
+
+struct SomeTag {};
+
+template <typename ExecutionSpace>
+struct TestRangePolicyConstructors {
+
+public:
+
+  TestRangePolicyConstructors() {
+    check_semiregular();
+    test_compile_time_parameters();
+    test_runtime_parameters();
+    test_worktag();
+    test_prefer_desired_occupancy();
+  }
+
+private:
+
+  void check_semiregular() {
+    // semiregular is copyable and default initializable
+    // (regular requires equality comparable)
+    using policy_t = Kokkos::RangePolicy<ExecutionSpace>;
+
+    static_assert(std::is_default_constructible<policy_t>::value, "");
+    static_assert(std::is_copy_constructible<policy_t>::value, "");
+    static_assert(std::is_move_constructible<policy_t>::value, "");
+    static_assert(std::is_copy_assignable<policy_t>::value, "");
+    static_assert(std::is_move_assignable<policy_t>::value, "");
+    static_assert(std::is_destructible<policy_t>::value, "");
+  }
+
+  void test_compile_time_parameters() {
+    {
+      using policy_t = Kokkos::RangePolicy<>;
+      using execution_space = typename policy_t::execution_space;
+      using index_type      = typename policy_t::index_type;
+      using schedule_type   = typename policy_t::schedule_type;
+      using work_tag        = typename policy_t::work_tag;
+
+      static_assert(std::is_same<execution_space, Kokkos::DefaultExecutionSpace>::value, "");
+      static_assert(std::is_same<index_type, typename execution_space::size_type>::value, "");
+      static_assert(std::is_same<schedule_type, Kokkos::Schedule<Kokkos::Static>>::value, "");
+      static_assert(std::is_void<work_tag>::value, "");
+    }
+
+    {
+      using policy_t = Kokkos::RangePolicy<ExecutionSpace>;
+      using execution_space = typename policy_t::execution_space;
+      using index_type      = typename policy_t::index_type;
+      using schedule_type   = typename policy_t::schedule_type;
+      using work_tag        = typename policy_t::work_tag;
+
+      static_assert(std::is_same<execution_space, ExecutionSpace>::value, "");
+      static_assert(std::is_same<index_type, typename execution_space::size_type>::value, "");
+      static_assert(std::is_same<schedule_type, Kokkos::Schedule<Kokkos::Static>>::value, "");
+      static_assert(std::is_void<work_tag>::value, "");
+    }
```

Comment:
If we are refactoring, let's improve readability.  Introduce a trait or a macro
```C++
CHECK_POLICY_PROPERTIES(Policy, ExecutionSpace, IndexType, WorkTag) \
static_assert(std::is_same_v(Policy::execution_space, ExecutionSpace); \
...
```
and then every block becomes a one-liner
```C++
CHECK_POLICY_PROPERTIES(Kokkos::RangePolicy<>,
  Kokkos::DefaultExecutionSpace, Kokkos::DefaultExecutionSpace::size_type, Kokkos::Schedule<Kokkos::Static>>::value, void);
CHECK_POLICY_PROPERTIES(Kokkos::RangePolicy<SomeTag>,
  Kokkos::DefaultExecutionSpace, Kokkos::DefaultExecutionSpace::size_type, Kokkos::Schedule<Kokkos::Static>>::value, SomeTag);
...

Reply:
I implemented something similar, but with a templated function. It is in `helper_functions/CommonExecPolicyTests.hpp`